engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ProgramExecutionSupport.scala

```diff
@@ -539,14 +546,41 @@ object ProgramExecutionSupport {
               .getOrElse(expressionValue.getClass)
           )
         )
-        ctx.executionService.callFunctionWithInstrument(
-          ctx.contextManager.getVisualizationHolder(contextId),
-          visualization.cache,
-          runtimeCache,
-          visualization.module,
-          visualization.callback,
-          expressionValue +: visualization.arguments: _*
-        )
+        val holder = ctx.contextManager.getVisualizationHolder(contextId)
+        val callback: Consumer[UUID] = if (runtimeCache != null) {
+          def processUUID(id: UUID): Unit = {
+            logger.log(
+              Level.WARNING,
+              "Associating visualization [{0}] with additional ID [{1}]",
+              Array[Object](
+                visualization.id,
+                id
+              )
+            )
+
+            holder.upsert(visualization, id)
```

Comment:
@4e6, I will need your help with the changes here. I don't fully understand how the `ProgramExecutionSupport`, `VisualizationHolder`, state, etc. work together. 

The idea is to collect all reads from the `RuntimeCache` when computing a visualization by `callFunctionWithInstrument`. This callback `upsert`s the visualization into the holder with UUID of the expression the visualization tries to read. When a value of any of such UUIDs changes, the visualization shall be re-evaluated _(that's not working yet for some reason)_.

Is this a sane idea? Is the implementation sound? Where is the best place to write a unit test for this behavior?

Reply:
> The idea is to collect all reads from the RuntimeCache

I don't think it is necessary to store all the values in the cache. The cache is required to store values between the runs. In this case, we only need to keep the value during a single program execution. I would introduce a `RuntimeState` that maintains the observable values. Then, from the runtime instrument, you can update the runtime state, and trigger the observables attached to it.

Reply:
I don't think there is any `RuntimeState` class. At least I cannot find one. Are you suggesting a refactoring? If so, then I'd rather avoid it until we agree the approach is correct.

>  we only need to keep the value during a single program execution

Right, that's why the values are put into the `RuntimeCache` as `WeakReference`s. Using `WeakReference` means, _be there_, but only until someone else _holds a reference to you_.

If the _value is temporary_ (e.g. no longer needed at the end of _instrumented execution_), then it must have been referenced only by the _single program execution_ stack frames. However stack frames are gone at the end of execution and as such **temporary values are eligible for GC** as soon as the execution is over.